stellargraph/layer/deep_graph_infomax.py

```diff
@@ -0,0 +1,156 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2020 Data61, CSIRO
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from . import GCN, GAT, APPNP, PPNP
+from ..core.experimental import experimental
+
+from tensorflow.keras.layers import Input, Lambda, Layer, GlobalAveragePooling1D
+import tensorflow as tf
+from tensorflow.keras import backend as K
+import warnings
+
+__all__ = ["DeepGraphInfoMax", "BilinearTransform"]
+
+
+class BilinearTransform(Layer):
```

Comment:
This is now a generic name that is a slightly peculiar bilinear transform, because it's not quite fully general like the PyTorch layer: that layer allows for `matrix * Weight * matrix + Bias` with pretty arbitrary dimensions, but this only allows `matrix * Weight * vector + 0` with `matrix.shape[1] == vector.shape[0]`. In addition, the documentation no longer quite matches the name: if it's going to be a bilinear transform, the docs should talk about it from that focus (possibly with a mention of DGI, but not only that).

To keep the scope small and thus keep this PR slightly simpler, I'd personally be happy to call this something very specific, e.g.

```suggestion
class DeepGraphInfoMaxDiscriminator(Layer):
```

or

```suggestion
class DGIDiscriminator(Layer):
```

and then, if we need it for other things in future, we can generalise and rename then (and it should be possible to have a migration path so that users of this layer directly aren't broken).

What do you think?

Reply:
I think that makes sense, as you mentioned its not quite a general bilinear transform and we can generalize it later if needed 👍